sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -312,13 +330,29 @@ export class BatchingReceiver extends MessageReceiver {
         }
       };
 
+      const cleanupBeforeReject = (
+        receiver: Receiver | undefined,
+        onReceiveErrorHandlerToRemove: OnAmqpEvent
+      ): void => {
+        if (receiver != null) {
+          receiver.removeListener(ReceiverEvents.receiverError, onReceiveErrorHandlerToRemove);
+          receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
+          receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
+          receiver.session.removeListener(SessionEvents.sessionError, onSessionError);
+        }
+
+        if (totalWaitTimer) {
+          clearTimeout(totalWaitTimer);
+        }
+        if (this._newMessageReceivedTimer) {
+          clearTimeout(this._newMessageReceivedTimer);
+        }
+      };
+
       // Action to be taken when an error is received.
-      const onReceiveError: OnAmqpEvent = (context: EventContext) => {
+      const onReceiveError: OnAmqpEvent = (context: Pick<EventContext, "receiver">) => {
         const receiver = this._receiver || context.receiver!;
-        receiver.removeListener(ReceiverEvents.receiverError, onReceiveError);
-        receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
-        receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
-        receiver.session.removeListener(SessionEvents.sessionError, onSessionError);
+        cleanupBeforeReject(receiver, onReceiveError);
```

Comment:
We can also reject from `sessionError`, it seems like this may be needed there as well.

Reply:
Well...it turns out this same code is basically in quite a few spots with slight variations. All unified, all now calling into this cleanup function (including a resolve code path).